### ISSUE TITLE
Pyjion 0.8.0 fails importing pickle with pyjion enabled

### ISSUE THREAD

**Issue by oroppas** [tonybaloney/Pyjion#166]
On Windows, `python -X dev -c "import pyjion; pyjion.enable(); import pickle"` randomly emits different errors:

```
C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'load_next_buffer' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in <module>
    class _Pickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in _Pickler
    class _Pickler:
NameError: name 'save_long' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in <module>
    class _Pickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in _Pickler
    class _Pickler:
NameError: name 'save_str' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'load_additems' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'SHORT_BINUNICODE' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in <module>
    class _Pickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 405, in _Pickler
    class _Pickler:
NameError: name 'save_picklebuffer' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1238, in _Unpickler
    dispatch[FRAME[0]] = load_frame
NameError: name 'FRAME' is not defined

C:\msys64\home\ryuta\packages\coolprop>python -X dev -c "import pyjion; pyjion.enable(); import pickle"
```

**Comment by tonybaloney** [comment]
Thanks for raising the issue. I wonder if this is caused by OPT-10. I'm also really confused why this isn't picked up in the smoke tests

**Comment by tonybaloney** [comment]
The name errors don't even match the code :-/

**Comment by tetsuo-cpp** [comment]
It might be platform-specific. I checked out the 0.8.0 commit but can't reproduce this on my MacBook.

**Comment by oroppas** [comment]
Trying on the REPL.

```
C:\Users\ryuta>python -X dev
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyjion
>>> pyjion.enable()
True
>>> import pickle
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'load_tuple3' is not defined
>>> quit()

C:\Users\ryuta>
C:\Users\ryuta>python -X dev
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyjion
>>> pyjion.enable()
True
>>> import pickle
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'load_long' is not defined
>>> quit()

C:\Users\ryuta>python -X dev
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyjion
>>> pyjion.enable()
True
>>> import pickle
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1134, in <module>
    class _Unpickler:
  File "C:\Users\ryuta\AppData\Local\Programs\Python\Python39\lib\pickle.py", line 1742, in _Unpickler
    dispatch[STOP[0]] = load_stop
NameError: name 'load_ext4' is not defined
>>> quit()
```

**Comment by tonybaloney** [comment]
@oroppas can you share `locale.localeconv()` and `locale.getconv()` on this ticket please.

**Comment by tonybaloney** [comment]
This can be reproduced by putting something which uses the OPT6/7 or OPT10 logic of precomputing hashes for either a subscript or a global. *But* it needs to be executed via `exec()`.
The only test which is failing in the smoke test has this same bug, to reproduce:

```python
    def test_bad_key(self):
        # Dictionary lookups should fail if __eq__() raises an exception.
        class CustomException(Exception):
            pass

        class BadDictKey:
            def __hash__(self):
                return hash(self.__class__)

            def __eq__(self, other):
                if isinstance(other, self.__class__):
                    raise CustomException
                return other

        d = {}
        x1 = BadDictKey()
        x2 = BadDictKey()
        d[x1] = 1
        for stmt in ['d[x2] = 2',
                     'z = d[x2]',
                     'x2 in d',
                     'd.get(x2)',
                     'd.setdefault(x2, 42)',
                     'd.pop(x2)',
                     'd.update({x2: 2})']:
            with self.assertRaises(CustomException):
                exec(stmt, locals())
```

To fix this, either:
* Find out what is special about hashing strings in this scenario and see if it can be fixed
* Disable this optimisation when called in this way (less preferred)

IIRC, this is PyRun_SimpleStringEx(), but that is the same logic used in the C++ unit tests, so I would've thought it would be reproduced there.

**Comment by tonybaloney** [comment]
@tetsuo-cpp this is a pretty juicy issue, do you want to have a go at this one next?

**Comment by tonybaloney** [comment]
Also worth noting that I've already written a `test_bad_key` with those tests above, but not through an exec() statement and it works fine.

See `Tests/test_dict.py`

**Comment by tetsuo-cpp** [comment]
> @tetsuo-cpp this is a pretty juicy issue, do you want to have a go at this one next?

Looks good. I'll have a go this week.

**Comment by oroppas** [comment]
```
>>> locale.localeconv()
{'int_curr_symbol': '', 'currency_symbol': '', 'mon_decimal_point': '', 'mon_thousands_sep': '', 'mon_grouping': [], 'positive_sign': '', 'negative_sign': '', 'int_frac_digits': 127, 'frac_digits': 127, 'p_cs_precedes': 127, 'p_sep_by_space': 127, 'n_cs_precedes': 127, 'n_sep_by_space': 127, 'p_sign_posn': 127, 'n_sign_posn': 127, 'decimal_point': '.', 'thousands_sep': '', 'grouping': []}
```

**Comment by tonybaloney** [comment]
> >>> locale.localeconv()
> {'int_curr_symbol': '', 'currency_symbol': '', 'mon_decimal_point': '', 'mon_thousands_sep': '', 'mon_grouping': [], 'positive_sign': '', 'negative_sign': '', 'int_frac_digits': 127, 'frac_digits': 127, 'p_cs_precedes': 127, 'p_sep_by_space': 127, 'n_cs_precedes': 127, 'n_sep_by_space': 127, 'p_sign_posn': 127, 'n_sign_posn': 127, 'decimal_point': '.', 'thousands_sep': '', 'grouping': []}

thanks, what is  `locale.getlocale()` and `locale.getdefaultlocale()`?

**Comment by oroppas** [comment]
Here we go:

```
>>> locale.getlocale()
('English_United States', 'utf8')
>>> locale.getdefaultlocale()
('en_US', 'cp65001')
```

**Comment by tonybaloney** [comment]
> Here we go:
> 
> ```
> >>> locale.getlocale()
> ('English_United States', 'utf8')
> >>> locale.getdefaultlocale()
> ('en_US', 'cp65001')
> ```

thanks @tetsuo-cpp I bet thats the cause ^^ see PEP538 https://www.python.org/dev/peps/pep-0538/

**Comment by tetsuo-cpp** [comment]
@tonybaloney 

I spent a bit of time on this today. I've narrowed it down to the `x2 in d` statement.

So this is fine:
```
with self.assertRaises(CustomException):
    x2 in d
```

But this isn't:
```
with self.assertRaises(CustomException):
    exec('x2 in d')
```

I've disassembled the first example and had a look at the way we handle the `CONTAIN_OP` opcode but it seems fine. One thing I wasn't sure about was how to inspect the bytecode generated by the `exec` call. Maybe that'd be helpful.

**Comment by tonybaloney** [comment]
> @tonybaloney
> 
> I spent a bit of time on this today. I've narrowed it down to the `x2 in d` statement.
> 
> So this is fine:
> 
> ```
> with self.assertRaises(CustomException):
>     x2 in d
> ```
> 
> But this isn't:
> 
> ```
> with self.assertRaises(CustomException):
>     exec('x2 in d')
> ```
> 
> I've disassembled the first example and had a look at the way we handle the `CONTAIN_OP` opcode but it seems fine. One thing I wasn't sure about was how to inspect the bytecode generated by the `exec` call. Maybe that'd be helpful.

Try disabling the optimisations in CMake config, e.g. `-DOPTIMIZE_HASHED_NAMES=OFF` and recompile.

**Comment by tetsuo-cpp** [comment]
> > @tonybaloney
> > I spent a bit of time on this today. I've narrowed it down to the `x2 in d` statement.
> > So this is fine:
> > ```
> > with self.assertRaises(CustomException):
> >     x2 in d
> > ```
> > 
> > 
> > But this isn't:
> > ```
> > with self.assertRaises(CustomException):
> >     exec('x2 in d')
> > ```
> > 
> > 
> > I've disassembled the first example and had a look at the way we handle the `CONTAIN_OP` opcode but it seems fine. One thing I wasn't sure about was how to inspect the bytecode generated by the `exec` call. Maybe that'd be helpful.
> 
> Try disabling the optimisations in CMake config, e.g. `-DOPTIMIZE_HASHED_NAMES=OFF` and recompile.

It still happens with `-DOPTIMIZE_HASHED_NAMES=OFF`. I'll try toggling the other optimizations and see if it makes a difference.

**Comment by tetsuo-cpp** [comment]
It seems to be failing even if I turn all the optimizations off in the `CMakeOptimizations.txt`.

@tonybaloney Do you know if there is a way to disassemble what gets generated by `eval`? I'm thinking that it'd be worth comparing the difference.

**Comment by tonybaloney** [comment]
Not to my knowledge, but I spotted this in the test suite (its the last passing test in the CI). Look at the file path. Its got some weird trailing unicode character at the end of it. Wondering if its related.


```
test_resource_tracker (test.test_multiprocessing_forkserver.TestResourceTracker) ... ok
38959
Traceback (most recent call last):
38960
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/multiprocessing/forkserver.py", line 274, in main
38961
    code = _serve_one(child_r, fds,
38962
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/multiprocessing/forkserver.py", line 313, in _serve_one
38963
    code = spawn._main(child_r, parent_sentinel)
38964
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
38965
    prepare(preparation_data)
38966
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
38967
    _fixup_main_from_path(data['init_main_from_path'])
38968
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
38969
    main_content = runpy.run_path(main_path,
38970
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/runpy.py", line 268, in run_path
38971
    return _run_module_code(code, init_globals, run_name,
38972
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/runpy.py", line 97, in _run_module_code
38973
    _run_code(code, mod_globals, init_globals,
38974
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/runpy.py", line 87, in _run_code
38975
    exec(code, run_globals)
38976
  File "/Users/runner/work/Pyjion/Pyjion/Tests/run_libregr_tests.py", line 6, in <module>
38977
    main()
38978
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/test/libregrtest/main.py", line 716, in main
38979
    Regrtest().main(tests=tests, **kwargs)
38980
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/test/libregrtest/main.py", line 638, in main
38981
    self._main(tests, kwargs)
38982
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/test/libregrtest/main.py", line 668, in _main
38983
    self.find_tests(tests)
38984
  File "/Users/runner/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/test/libregrtest/main.py", line 218, in find_tests
38985
    with open(os.path.join(support.SAVEDCWD, self.ns.fromfile)) as fp:
38986
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/test_python_2934æ/Tests/python_tests_macos10.txt'
```

**Comment by tonybaloney** [comment]
This is the test I've added to reproduce it in C++

```cpp
TEST_CASE("Test dict hashes") {
    SECTION("test precomputed hash on dict") {
        auto t = CompilerTest(
                "def f():\n"
                "    l = {'a': 1, 'b': 2}\n"
                "    l['a'] = 3\n"
                "    return l['a']"
        );
        CHECK(t.returns() == "3");
    }
    SECTION("test precomputed hash on dict within exec") {
        auto t = CompilerTest(
                "def f():\n"
                "    l = {'a': 1, 'b': 2}\n"
                "    exec('l[\"a\"] = 3')\n"
                "    return l['a']\n"
        );
        CHECK(t.returns() == "3");
    }
}
```

**Comment by tonybaloney** [comment]
output: 

```default
Testing started at 2:28 pm ...
Traceback (most recent call last):
  File "<string>", line 3, in f
  File "<string>", line 1, in <module>
NameError: name 'l' is not defined
Collected 0 values
load name failed
call function failed
Collected 0 values

/Users/anthonyshaw/CLionProjects/pyjion/Tests/test_compiler.cpp:75: Failure:
  REQUIRE(res.get() != nullptr)
with expansion:
  nullptr != nullptr

/Users/anthonyshaw/CLionProjects/pyjion/Tests/test_compiler.cpp:1690: Failure:
  CHECK(t.returns() == "3")
with expansion:
  t.returns() == "3"
due to unexpected exception with message:
  Exception translation was disabled by CATCH_CONFIG_FAST_COMPILE


Process finished with exit code 2
```

**Comment by tonybaloney** [comment]
The operation that fails is `LOAD_NAME`. That calls `PyJit_LoadName` or `PyJit_LoadNameHash` (if `DOPTIMIZE_HASHED_NAMES=ON`).

I put a breakpoint in and weirdly, `f_locals` is an empty dictionary.

```cpp
PyObject* PyJit_LoadName(PyFrameObject* f, PyObject* name) {
    PyObject *locals = f->f_locals;
```

So this would fail with or without the optimisation.

**Comment by tonybaloney** [comment]
Short-term workaround https://github.com/tonybaloney/Pyjion/pull/205

**Comment by tetsuo-cpp** [comment]
@tonybaloney
Sorry for disappearing.
Does this need more investigation? Or do you expect to #206 to solve it.

**Comment by nlhnt** [comment]
Just found this issue, and I think what I posted is connected.  
https://github.com/tonybaloney/Pyjion/issues/301

**Comment by tonybaloney** [comment]
resolved in 205, #301 is not related.
